File: app/src/main/java/com/blog/demo/application/FilterActivity.java
package com.blog.demo.application;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Filter;
import android.widget.ListView;

import com.blog.demo.R;

import java.util.ArrayList;
import java.util.List;

public class FilterActivity extends Activity {

    private EditText mEtInput;
    private ArrayAdapter mAdapter;
    private Filter mFilter = new SearchFilter();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_application_filter);

        mEtInput = findViewById(R.id.et_input);
        ListView listView = findViewById(R.id.list_view);
        mAdapter = new ArrayAdapter<String>(this, R.layout.list_view_item, R.id.tv_name);
        listView.setAdapter(mAdapter);

        mEtInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                mFilter.filter(mEtInput.getText().toString());
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private class SearchFilter extends Filter {
        private List<String> mTextList = new ArrayList<>();

        public SearchFilter() {
            mTextList.add("abc");
            mTextList.add("abd");
            mTextList.add("abe");
            mTextList.add("acd");
            mTextList.add("acde");
            mTextList.add("ace");
            mTextList.add("bcd");
            mTextList.add("bce");
            mTextList.add("bcf");
        }

        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            // FilterResults用来保存过滤结果，包含两个属性values和count
            FilterResults results = new FilterResults();

            List<String> resultList = new ArrayList<>();
            for (String text : mTextList) {
                if (constraint != null && constraint.length() > 0 && text.contains(constraint)) {
                    resultList.add(text);
                }
            }

            results.values = resultList;
            results.count = resultList.size();

            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            mAdapter.clear();
            if (results.count > 0) {
                mAdapter.addAll((List<String>)results.values);
            }
        }
    }

}

File: app/src/main/java/com/blog/demo/feature/FilterActivity.kt
package com.blog.demo.feature

import android.app.Activity
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Filter
import android.widget.ListView
import com.blog.demo.R

class FilterActivity : Activity() {
    private lateinit var mEtInput: EditText
    private lateinit var mAdapter: ArrayAdapter<String>
    private val mFilter: Filter = SearchFilter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_feature_filter)
        mEtInput = findViewById(R.id.et_input)

        val listView: ListView = findViewById(R.id.list_view)
        mAdapter = ArrayAdapter<String>(this, R.layout.list_view_item, R.id.tv_name)
        listView.adapter = mAdapter

        mEtInput.addTextChangedListener(object : TextWatcher {

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                mFilter.filter(mEtInput.text.toString())
            }

            override fun afterTextChanged(s: Editable) {}

        })
    }

    private inner class SearchFilter : Filter() {
        private val mTextList: MutableList<String> = ArrayList()

        init {
            mTextList.add("abc")
            mTextList.add("abd")
            mTextList.add("abe")
            mTextList.add("acd")
            mTextList.add("acde")
            mTextList.add("ace")
            mTextList.add("bcd")
            mTextList.add("bce")
            mTextList.add("bcf")
        }

        override fun performFiltering(constraint: CharSequence): FilterResults {
            // FilterResults用来保存过滤结果，包含两个属性values和count
            val results = FilterResults()
            val resultList: MutableList<String> = ArrayList()
            for (text in mTextList) {
                if (constraint.isNotEmpty() && text.contains(constraint)) {
                    resultList.add(text)
                }
            }
            results.values = resultList
            results.count = resultList.size

            return results
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            mAdapter.clear()
            if (results.count > 0) {
                mAdapter.addAll(results.values as List<String>)
            }
        }

    }

}